go.mod

```diff
@@ -1,76 +1,40 @@
-	github.com/gogo/protobuf v1.0.0
 	github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
-	github.com/golang/groupcache v0.0.0-20180513044358-24b0969c4cb7
-	github.com/golang/protobuf v1.1.0
-	github.com/google/btree v0.0.0-20180124185431-e89373fe6b4a
-	github.com/google/gofuzz v0.0.0-20170612174753-24818f796faf
-	github.com/googleapis/gnostic v0.2.0
-	github.com/gregjones/httpcache v0.0.0-20180305231024-9cad4c3443a7
-	github.com/hashicorp/golang-lru v0.0.0-20180201235237-0fb14efe8c47
-	github.com/howeyc/gopass v0.0.0-20170109162249-bf9dde6d0d2c
-	github.com/imdario/mergo v0.3.5
-	github.com/inconshreveable/mousetrap v1.0.0
-	github.com/json-iterator/go v0.0.0-20180612202835-f2b4162afba3
-	github.com/mailru/easyjson v0.0.0-20180606163543-3fdea8d05856
-	github.com/matttproud/golang_protobuf_extensions v1.0.1
-	github.com/mitchellh/go-wordwrap v0.0.0-20150314170334-ad45545899c7
-	github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
-	github.com/modern-go/reflect2 v0.0.0-20180228065516-1df9eeb2bb81
-	github.com/opencontainers/go-digest v1.0.0-rc1
-	github.com/pborman/uuid v0.0.0-20170612153648-e790cca94e6c
-	github.com/petar/GoLLRB v0.0.0-20130427215148-53be0d36a84c
-	github.com/peterbourgon/diskv v2.0.1+incompatible
-	github.com/pmezard/go-difflib v1.0.0
-	github.com/prometheus/client_golang v0.8.0
-	github.com/prometheus/client_model v0.0.0-20171117100541-99fa1f4be8e5
-	github.com/prometheus/common v0.0.0-20180518154759-7600349dcfe1
-	github.com/prometheus/procfs v0.0.0-20180612222113-7d6f385de8be
-	github.com/russross/blackfriday v0.0.0-20151117072312-300106c228d5
-	github.com/shurcooL/sanitized_anchor_name v0.0.0-20170918181015-86672fcb3f95
-	github.com/spf13/cobra v0.0.3
-	github.com/spf13/pflag v1.0.1
-	github.com/stretchr/testify v1.2.2
-	golang.org/x/crypto v0.0.0-20180621125126-a49355c7e3f8
-	golang.org/x/net v0.0.0-20180629035331-4cb1c02c05b0
-	golang.org/x/sys v0.0.0-20180627142611-7138fd3d9dc8
-	golang.org/x/text v0.3.0
-	golang.org/x/time v0.0.0-20180412165947-fbb02b2291d2
-	gopkg.in/inf.v0 v0.9.1
-	gopkg.in/square/go-jose.v2 v2.1.6
-	gopkg.in/yaml.v2 v2.2.1
-	k8s.io/api v0.0.0-20180621150657-6c0bbc3e58fa
-	k8s.io/apiextensions-apiserver v0.0.0-20180627192848-4340dd9c2406
-	k8s.io/apimachinery v0.0.0-20180619225948-e386b2658ed2
-	k8s.io/apiserver v0.0.0-20180621154834-1bfbd2d59262
-	k8s.io/autoscaler v0.0.0-20180517133234-a909d0cc3695
-	k8s.io/client-go v7.0.0+incompatible
-	k8s.io/kube-openapi v0.0.0-20180629012420-d83b052f768a
-	k8s.io/kubernetes v1.10.5
-	k8s.io/utils v0.0.0-20180621173828-733eca437aa3
-	vbom.ml/util v0.0.0-20170409195630-256737ac55c4
+	github.com/prometheus/client_golang v1.7.1
+	github.com/spf13/pflag v1.0.5
+	github.com/stretchr/testify v1.4.0
+	k8s.io/api v0.19.2
+	k8s.io/apimachinery v0.19.2
+	k8s.io/autoscaler/cluster-autoscaler v0.0.0-20200917024141-03f60a4c3818
+	k8s.io/client-go v0.19.2
+	k8s.io/kubernetes v0.0.0
+)
+
+replace (
+	k8s.io/api => k8s.io/api v0.19.2
```

Comment:
Why using replace statement here?

Reply:
Great question too! 

This project was not done using Go Mod, but rather Glide initially. A lot of the direct dependencies are also not go mods themselves. Removing this leads to those dependencies pulling different versions of transient dependencies, especially Kubernetes ones. This forces the versions of those specific dependencies. Some of those direct dependencies declared direct dependence on kubernetes, so did the k8s-spot-rescheduler for access to the legacyscheme versions of the api. Googling the errors on dependencies resolution eventually lead me here: https://github.com/kubernetes/kubernetes/issues/79384 And this seemed to be the main way to guarantee those versions.

Reply:
Interesting answer. I am not sure to understand the reason for each of them though.

Reply:
It is more of a package deal really, for Kubernetes related dependencies. In the github issue mentioned there is a script to help generate that block for a specific version of kubernetes. I went for 1.19.2 and selected the version of cluster-autoscaler that depended on that version. They are really quick to adapt, and already have code relying on recent alpha features in RC newer than that, and early builds of 1.19.3, so I really picked the latest they had for 1.19.2. 

I'll go through and remove some of the more outlandish overrides and run test coverage and live debugging, see if I can trim the list down.

Reply:
Ok @jocgir few more things done. I removed the direct dependency on `k8s.io/kubernetes` that was present directly in k8s-spot-rescheduler. The dependency on legacy scheme was easily replaced. Tested with `go test --cover ./...` and a direct scaling and rescheduling on dev eu-west-1 mock-dpm test run. 

However, the cluster-autoscaler also depends directly on that package, and one by one, I had to put back all the replaces.. I think as a result of this comment, the direct project got better and more compliant, but the direct dependency of the cluster autoscaler didn't, so we need that replace block, and every line of it too.

Reply:
Would be better to stick to deployed kubernetes Version. You can use 1.17 as update is coming soon.